hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/MaxRateMetrics.java

```diff
@@ -0,0 +1,63 @@
+import static pl.allegro.tech.hermes.common.metric.Counters.MAXRATE_FETCH_FAILURES;
+import static pl.allegro.tech.hermes.common.metric.Counters.MAXRATE_RATE_HISTORY_FAILURES;
+import static pl.allegro.tech.hermes.common.metric.Gauges.MAX_RATE_ACTUAL_RATE_VALUE;
+import static pl.allegro.tech.hermes.common.metric.Gauges.MAX_RATE_CALCULATION_DURATION;
+import static pl.allegro.tech.hermes.common.metric.Gauges.MAX_RATE_VALUE;
+import static pl.allegro.tech.hermes.common.metric.Gauges.OUTPUT_RATE;
+import static pl.allegro.tech.hermes.common.metric.SubscriptionTagsFactory.subscriptionTags;
+
+public class MaxRateMetrics {
+
+    private final HermesMetrics hermesMetrics;
+    private final MeterRegistry meterRegistry;
+
+    MaxRateMetrics(HermesMetrics hermesMetrics, MeterRegistry meterRegistry) {
+        this.hermesMetrics = hermesMetrics;
+        this.meterRegistry = meterRegistry;
+    }
+
+    public <T> void registerCalculationDurationInMillisGauge(T obj, ToDoubleFunction<T> f) {
+        hermesMetrics.registerGauge(MAX_RATE_CALCULATION_DURATION, () -> (int) f.applyAsDouble(obj));
+        meterRegistry.more().timeGauge("max-rate.calculation.duration", List.of(), obj, MILLISECONDS, f);
+    }
+
+    public HermesCounter historyUpdateFailuresCounter(SubscriptionName subscription) {
+        return HermesCounters.from(
+                meterRegistry.counter("max-rate.history-update.failures", subscriptionTags(subscription)),
+                hermesMetrics.counter(MAXRATE_RATE_HISTORY_FAILURES, subscription.getTopicName(), subscription.getName())
+        );
+    }
+
+    public HermesCounter fetchFailuresCounter(SubscriptionName subscription) {
+        return HermesCounters.from(
+                meterRegistry.counter("max-rate.fetch.failures", subscriptionTags(subscription)),
+                hermesMetrics.counter(MAXRATE_FETCH_FAILURES, subscription.getTopicName(), subscription.getName())
+        );
+    }
+
+    public <T> void registerCalculatedRateGauge(SubscriptionName subscription, T obj, ToDoubleFunction<T> f) {
+        hermesMetrics.registerGauge(MAX_RATE_VALUE, subscription, () -> (int) f.applyAsDouble(obj));
+        meterRegistry.gauge("max-rate.calculated-rate", subscriptionTags(subscription), obj, f);
+    }
+
+    public <T> void registerActualRateGauge(SubscriptionName subscription, T obj, ToDoubleFunction<T> f) {
+        hermesMetrics.registerGauge(MAX_RATE_ACTUAL_RATE_VALUE, subscription, () -> (int) f.applyAsDouble(obj));
+        meterRegistry.gauge("max-rate.actual-rate", subscriptionTags(subscription), obj, f);
+    }
+
+    public <T> void registerOutputRateGauge(SubscriptionName subscription, T obj, ToDoubleFunction<T> f) {
+        hermesMetrics.registerGauge(OUTPUT_RATE, subscription, () -> (int) f.applyAsDouble(obj));
+        // This metric provides almost the same information as Gauge.MAX_RATE_ACTUAL_RATE_VALUE does, so there is no need to migrate it to Micrometer.
```

Comment:
Good point. I made a mistake in the comment. I meant `Gauge.MAX_RATE_VALUE`, not `Gauge.MAX_RATE_ACTUAL_RATE_VALUE`. Please, double-check if this is correct now.

Reply:
If I understand correctly:

OUTPUT_RATE - the rate set in RateLimiter
MAX_RATE_VALUE - rate calculated by NegotiatedMaxRateProvider
MAX_RATE_ACTUAL_RATE_VALUE - the actual observed rate

From what I can see MAX_RATE_VALUE can be modified by ModeOutputRateCalculator so the OUTPUT_RATE can be different than MAX_RATE_VALUE. Are we sure that OUTPUT_RATE is not valuable on its own? It is as close to rate limiter as it can be so it can useful for debugging, what do you think?

Reply:
It looks like you're right. I don't know why, but I have the impression that I've never noticed any differences in charts built upon these metrics in production. This is strange, especially considering that the heartbeat mode should cause such differences. Anyway, I've migrated the OUTPUT_RATE metric to Micrometer. 

Thanks for pointing this out!